semantic-analysis/src/Analysis/Analysis/Exception.hs

```diff
@@ -28,91 +36,158 @@ import           Analysis.File
-fromExceptions = ExcSet mempty . Set.fromList . Foldable.toList
+fromExceptions es = ExcSet mempty (Set.fromList (Foldable.toList es)) mempty
 
-var :: Name -> ExcSet
-var v = ExcSet (Set.singleton v) mempty
+var :: FreeVariable -> ExcSet
+var v = ExcSet (Set.singleton v) mempty mempty
 
 exc :: Exception -> ExcSet
-exc e = ExcSet mempty (Set.singleton e)
+exc e = ExcSet mempty (Set.singleton e) mempty
+
+str :: Text.Text -> ExcSet
+str s = ExcSet mempty mempty (Set.singleton s)
 
 subst  :: Name -> ExcSet -> ExcSet -> ExcSet
-subst name (ExcSet fvs' es') (ExcSet fvs es) = ExcSet (Set.delete name fvs <> fvs') (es <> es')
+-- FIXME: this doesn't handle transitivity at all.
+subst name (ExcSet _ es' _) (ExcSet fvs es ss) = ExcSet fvs'' (es <> es'') ss
+  where
+  (fvs'', es'') = foldMap combine fvs
+  combine fv
+    | freeVariableName fv == name = (mempty, Set.map (\ (Exception n _) -> Exception n (freeVariableLines fv)) es')
+    | otherwise                   = (Set.singleton fv, mempty)
+
+
+nullExcSet :: ExcSet -> Bool
+nullExcSet e = null (freeVariables e) && null (exceptions e)
+
+freeVariablesForLine :: Int -> ExcSet -> Set.Set FreeVariable
+freeVariablesForLine l e = Set.filter (\ fv -> IntSet.member l (freeVariableLines fv)) (freeVariables e)
+
+exceptionsForLine :: Int -> ExcSet -> Set.Set Exception
+exceptionsForLine l e = Set.filter (\ ex -> IntSet.member l (exceptionLines ex)) (exceptions e)
 
+printExcSet :: Source.Source -> ExcSet -> IO ()
+printExcSet src e = for_ (zip [0..] (Source.lines src)) $ \ (i, line) -> do
+  Text.putStr (keywords (Text.dropWhileEnd (== '\n') (Source.toText line)))
+  let es  = exceptionsForLine    i e
+      fvs = freeVariablesForLine i e
+  unless (null es && null fvs) $ do
+    Text.putStr (Text.pack " \ESC[30;1m# ")
+    Text.putStr (Text.pack "{" <> union
+      (  formatFreeVariables fvs
+      <> formatExceptions    es ) <> Text.pack "}" <> reset)
+  Text.putStrLn mempty
+  where
+  keyword k s = Text.intercalate (Text.pack "\ESC[34;1m" <> k <> reset) (Text.splitOn k s)
+  keywords = keyword (Text.pack "raise") . keyword (Text.pack "import") . keyword (Text.pack "def") . keyword (Text.pack "pass")
+  union = Text.intercalate (Text.pack ", ")
+  formatFreeVariables fvs  = map (\ fv -> Text.pack "?" <> formatName (freeVariableName fv)) (Set.toList fvs)
+  formatExceptions    excs = map (Text.pack . show . formatName . exceptionName) (Set.toList excs)
+  reset = Text.pack "\ESC[0m"
+
+refLines :: Reference -> IntSet.IntSet
+refLines (Reference _ (Span (Pos startLine _) (Pos endLine _))) = IntSet.fromAscList [startLine..endLine]
 
 exceptionTracing
   :: Ord term
   => ( forall sig m
-     .  (Has (Env A.MAddr) sig m, HasLabelled Store (Store A.MAddr ExcSet) sig m, Has (Dom ExcSet) sig m, Has A.Statement sig m)
+     .  (Has (Env A.MAddr) sig m, HasLabelled Store (Store A.MAddr ExcSet) sig m, Has (Dom ExcSet) sig m, Has (Reader Reference) sig m, Has A.Statement sig m)
```

Comment:
We now carry around a `Reference` so we can track the current path and span as we go.